Add ecobee ventilator

marcolivierarsenault opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Adding a new number entity to ecobee thermostat to represent the ventilator (also called air exchanger or hrv). This new entity allow you to set runtime (min/hour) for home or away setting.

This PR is a remake of that old close pr #51800



## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/25215

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

marcolivierarsenault committed:
add ecobee ventilator

marcolivierarsenault committed:
add testing

gjohansson-ST commented on homeassistant/components/ecobee/const.py:
```diff
@@ -46,6 +46,7 @@
     Platform.HUMIDIFIER,
     Platform.SENSOR,
     Platform.WEATHER,
+    Platform.FAN,
```
I think it should be good practice to set them in alphabetic order

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(minutes=3)
```
```suggestion
```
No usage when coming from a config entry

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+
+    async def async_update(self):
+        """Get the latest state from the thermostat."""
+        if self.update_without_throttle:
```
Looks like it will never be true so could be removed completely?

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information for the ecobee ventilator."""
+        model: str | None
+        try:
+            model = f"{ECOBEE_MODEL_TO_NAME[self.thermostat['modelNumber']]} Thermostat"
+        except KeyError:
+            # Ecobee model is not in our list
+            model = None
```
```suggestion
        model = ECOBEE_MODEL_TO_NAME.get(self.thermostat['modelNumber'])
```

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+            # Ecobee model is not in our list
+            model = None
+
+        return DeviceInfo(
```
With the change above you can move this to init with self._attr_device_info

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+        self.thermostat_index = thermostat_index
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
+        ecobee_name = self.thermostat["name"]
+        self._name = f"{ecobee_name} Ventilator"
```
```suggestion
        self._attr_name= f"{ecobee_name} Ventilator"
```

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+
+        self.update_without_throttle = False
+
+    @property
+    def name(self):
+        """Return the name of the ventilator."""
+        return self._name
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,187 @@
+    @property
+    def unique_id(self):
+        """Return unique_id for ventilator."""
+        return f"{self.thermostat['identifier']}"
```
Move to init with self._attr_unique_id

gjohansson-ST commented on .coveragerc:
```diff
@@ -259,6 +259,7 @@ omit =
     homeassistant/components/ecobee/__init__.py
     homeassistant/components/ecobee/binary_sensor.py
     homeassistant/components/ecobee/climate.py
+    homeassistant/components/ecobee/fan.py
```
Why is it added here when you have implemented tests?

marcolivierarsenault replied:
@gjohansson-ST  , the test does not provide 100% coverage, I did not find the way to test the [async_setup_entry](https://github.com/home-assistant/core/pull/83645/files#diff-a512bf42b37bf296aa4c7a4460458217d634f473106d25f58cb7dd0acc4b4553R33) so I added the file to `.coveragerc` (This is also inline with other Ecobee entities (climate, sensors, etc))

Let me know if there is a better way to do it.

gjohansson-ST replied:
You need to mock a config entry and use that to setup the integration.
Search the codebase for `MockConfigEntry` to see multiple examples or here is one: https://github.com/home-assistant/core/blob/dev/tests/components/sensibo/conftest.py#L23

See if you can make it to not add it to coverage for exclusion.

marcolivierarsenault replied:
Looking at this, Thank a lot

gjohansson-ST reviewed (changes requested):
Some review comments

marcolivierarsenault committed:
Update homeassistant/components/ecobee/fan.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

marcolivierarsenault committed:
Update homeassistant/components/ecobee/fan.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

marcolivierarsenault committed:
Update homeassistant/components/ecobee/fan.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

marcolivierarsenault committed:
Update homeassistant/components/ecobee/fan.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

marcolivierarsenault committed:
move properties to init, apply pr review comments

marcolivierarsenault committed:
add identifier in test fixture

marcolivierarsenault committed:
add 100% codecov

marcolivierarsenault commented:
@gjohansson-ST Thanks a ton for your feedback:
- Comments applied
- Removed file from `.coveragerc`
- Adjusted testing to reach 100%

Name                                             Stmts   Miss  Cover   Missing
------------------------------------------------------------------------------
homeassistant/components/ecobee/fan.py              65      0   100%

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+        "set_ventilator_min_on_time_away",
+    )
+
+    platform.async_register_entity_service(
```
Looks like it should be implemented as a `switch` instead of a custom service?

marcolivierarsenault replied:
removed

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
```
Looks like it can be implemented using the `number` platform instead of custom service. Same with the one below.

marcolivierarsenault replied:
@gjohansson-ST I thought about this but it does not seems to support `time` or `minutes` as a unit: https://developers.home-assistant.io/docs/core/entity/number#available-device-classes

gjohansson-ST replied:
It doesn't. Maybe you're right. It looks as there is a `time` platform coming so maybe do a custom service now and transition to the new platform when it's launched.

marcolivierarsenault replied:
Kept it as a separate function, I will keep an eye on future time platform and update the integration

frenck replied:
The `time` platform in progress is not for `duration`. Time represents a time of day, which is not a fit for this.

This should be a `number` entity.

marcolivierarsenault replied:
Thanks @frenck 
Happy to turn those 2 into a `number` entity.

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+        self.data = data
+        self.thermostat_index = thermostat_index
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
+        ecobee_name = self.thermostat["name"]
+        self._attr_name = f"{ecobee_name} Ventilator"
```
```suggestion
        self._attr_name = f"{self.thermostat["name"]} Ventilator"
```

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+            identifiers={(DOMAIN, self.thermostat["identifier"])},
+            manufacturer=MANUFACTURER,
+            model=ECOBEE_MODEL_TO_NAME.get(self.thermostat["modelNumber"]),
+            name=self.name,
```
```suggestion
            name=self._attr_name,
```
You might want to implement [this](https://developers.home-assistant.io/docs/core/entity?_highlight=has_en#has_entity_name-true-mandatory-for-new-integrations) right away

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+            model=ECOBEE_MODEL_TO_NAME.get(self.thermostat["modelNumber"]),
+            name=self.name,
+        )
+        self._attr_unique_id = f"{self.thermostat['identifier']}"
```
```suggestion
        self._attr_unique_id = self.thermostat['identifier']
```

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+        )
+        self._attr_unique_id = f"{self.thermostat['identifier']}"
+
+        self.update_without_throttle = False
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the ventilator 20 min manual timer."""
```
Shouldn't these be manual e.g. no timer?

marcolivierarsenault replied:
This type of ventilator does not offer manual on/off only via timer and time per minutes on home or away

gjohansson-ST reviewed (changes requested):
Some additional comments/questions.
You should probably type all of it as well.

marcolivierarsenault committed:
apply pr comments

marcolivierarsenault committed:
remove custom service for timer

marcolivierarsenault commented on homeassistant/components/ecobee/fan.py:
```diff
@@ -0,0 +1,161 @@
+            ],
+            "is_ventilator_timer_on": self.thermostat["settings"][
+                "isVentilatorTimerOn"
+            ],
```
Kept Custom services, for the 2 timers. 
Removed the timer on

marcolivierarsenault commented:
@gjohansson-ST 
Applied your comments. 

Removed the timer custom-service. kept the two timers seems it is still, what I think, the most appropriate entity until we have Time Platform. 

On/Off with the timer (20 min) since this is how the actual physical device does work, there is no real on/off it is instead by manual 20 min show.

marcolivierarsenault committed:
move min time to number entities

marcolivierarsenault commented:
@frenck and @gjohansson-ST 
I applied your comments and move the 2 custom services to `number` entity.

Thanks for the help

marcolivierarsenault committed:
remove custom service definition

marcolivierarsenault committed:
generalize number class

gjohansson-ST commented:
Typically implement one platform per PR so I suggest you move the implementation of `number` to it's own PR

marcolivierarsenault commented:
@gjohansson-ST any objections as I keep number and open a pr for the fan. 

My priorities is with those 2 numbers. The manual fan is bonus.

marcolivierarsenault committed:
Move fan entity to new commit

marcolivierarsenault commented:
@gjohansson-ST done

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+        """Initialize ecobee ventilator platform."""
+        self.data = data
+        self.thermostat_index = thermostat_index
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
```
```suggestion
        self.thermostat = self.data.ecobee.get_thermostat(thermostat_index)
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+
+class EcobeeVentilatorMinTime(NumberEntity):
+    """A number class, representing min time  for an ecobee thermostat with ventilator attached."""
+
```
```suggestion

    _attr_has_entity_name = True
    
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
+        self.ecobee_setting_name = ecobee_setting_name
+        self.set_func = set_func
+        self._attr_has_entity_name = True
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+class EcobeeVentilatorMinTime(NumberEntity):
+    """A number class, representing min time  for an ecobee thermostat with ventilator attached."""
+
+    def __init__(self, data, thermostat_index, name, ecobee_setting_name, set_func):
```
is `name` actually `mode`? For me "home" doesn't neccessarily represent a name?

gjohansson-ST commented on tests/components/ecobee/test_number.py:
```diff
@@ -0,0 +1,95 @@
+
+
+@pytest.fixture(name="number")
+def home_number_fixture(data):
```
I think we should test by setting up the integration

marcolivierarsenault replied:
This is already tested in the [integration unit tests](https://github.com/home-assistant/core/blob/d9903c4cf985381002de8b923815b05dd24e0651/tests/components/ecobee/common.py#L11)

By adding the setting info in the fixture
```yaml
        "ventilatorType": "hrv",
        "ventilatorMinOnTimeHome": 20,
        "ventilatorMinOnTimeAway": 10,
        "isVentilatorTimerOn": false,
```
https://github.com/home-assistant/core/pull/83645/files#diff-1c4cbfd307eeb9092402c30e437b489816d22c104420d64f1e929026d82f300eR30

we do test the addition of the numbers

marcolivierarsenault replied:
ooo nevermind, I understand what you mean. 

I applied your feedback

gjohansson-ST reviewed (changes requested):
Some cleanup comments + I think we should test by setting up the integration.

marcolivierarsenault committed:
Update homeassistant/components/ecobee/number.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

marcolivierarsenault committed:
add cleanup comments

marcolivierarsenault commented:
@gjohansson-ST Reviewed and applied your comment. 

Added details about integration test being done already by the overall Ecobee integration.

marcolivierarsenault committed:
change test to integration setup

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,94 @@
+            name=self.thermostat["name"],
+        )
+        self._attr_unique_id = f'{self.thermostat["identifier"]}_{mode}'
+        self._attr_native_min_value = 0
+        self._attr_native_max_value = 60
+        self._attr_native_step = 5
```
Static values should be set at class level

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,94 @@
+        self._attr_native_max_value = 60
+        self._attr_native_step = 5
+        self._attr_native_value = self.thermostat["settings"][ecobee_setting_key]
+        self._attr_native_unit_of_measurement = UnitOfTime.MINUTES
```
as above

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,94 @@
+    for index in range(len(data.ecobee.thermostats)):
+        thermostat = data.ecobee.get_thermostat(index)
+        if thermostat["settings"]["ventilatorType"] != "none":
+            _LOGGER.debug("Adding 1 ventilator's min time home number")
```
You should probably expand the text to include the ventilator so someone debugging get's some more input

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,94 @@
+            model=ECOBEE_MODEL_TO_NAME.get(self.thermostat["modelNumber"]),
+            name=self.thermostat["name"],
+        )
+        self._attr_unique_id = f'{self.thermostat["identifier"]}_{mode}'
```
Seems a bit week to only add the mode to the identifier?

marcolivierarsenault replied:
Not sure if you think of a better way, 
technically each ecobee device can only have 1 ventilator, and only 2 move (home and away)

Pairing mode with thermostat name (aka device name) makes it unique

gjohansson-ST replied:
In that case we can leave it

gjohansson-ST replied:
Probably you should add `ventilator` to the unique_id if somewhere there are other modes coming in not seen today to differentiate it.

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,94 @@
+
+    def set_native_value(self, value: float) -> None:
+        """Set new ventilator Min On Time value."""
+        self._attr_native_value = int(value)
```
This line has no effect unless you also write the changes to the state machine.
But then you should also verify the function was successful at first.

gjohansson-ST commented on tests/components/ecobee/test_number.py:
```diff
@@ -0,0 +1,70 @@
+    assert state.attributes.get("unit_of_measurement") == UnitOfTime.MINUTES
+
+
+async def test_set_min_time_home(hass):
```
I think this test is a bit week.
I think you should setup the platform and assert the state, then call the service and assert the state has changed to what you set in.

marcolivierarsenault replied:
Right now, in the code, I am waiting for a updated value from the web service to update the value. (aka, I don't store it locally) Should I do it differently?

gjohansson-ST replied:
You can monkeypatch the fixture to alter the value and then move time so the update runs and get's the new number.
Example: https://github.com/home-assistant/core/blob/dev/tests/components/sensibo/test_number.py#L39-L49

marcolivierarsenault replied:
ok got it, done

gjohansson-ST reviewed (changes requested):
Last bits and pieces from me I think.

marcolivierarsenault committed:
apply comments

marcolivierarsenault commented:
Thanks @gjohansson-ST I applied most of your comments, left some comment regarding 2 of them

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,98 @@
+
+class EcobeeVentilatorMinTime(NumberEntity):
+    """A number class, representing min time  for an ecobee thermostat with ventilator attached."""
+
```
```suggestion
    _attr_native_min_value = VENTILATOR_MIN_VALUE
    _attr_native_max_value = VENTILATOR_MAX_VALUE
    _attr_native_step = VENTILATOR_STEP
    _attr_native_unit_of_measurement = UnitOfTime.MINUTES
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,98 @@
+            name=self.thermostat["name"],
+        )
+        self._attr_unique_id = f'{self.thermostat["identifier"]}_{mode}'
+        self._attr_native_min_value = VENTILATOR_MIN_VALUE
+        self._attr_native_max_value = VENTILATOR_MAX_VALUE
+        self._attr_native_step = VENTILATOR_STEP
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,98 @@
+        self._attr_native_max_value = VENTILATOR_MAX_VALUE
+        self._attr_native_step = VENTILATOR_STEP
+        self._attr_native_value = self.thermostat["settings"][ecobee_setting_key]
+        self._attr_native_unit_of_measurement = VENTILATOR_MEASUREMENT_UNIT
```
```suggestion
```

marcolivierarsenault replied:
Why did you remove the `native_unit_of_measurement`

gjohansson-ST replied:
Was meant to go to class level as it's a fixed record.

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,98 @@
+VENTILATOR_MIN_VALUE = 0
+VENTILATOR_MAX_VALUE = 60
+VENTILATOR_STEP = 5
+VENTILATOR_MEASUREMENT_UNIT = UnitOfTime.MINUTES
```
```suggestion
```

gjohansson-ST reviewed (changes requested):
Perhaps you misunderstood my comment on moving to class level

marcolivierarsenault committed:
move class value, add mock testing

marcolivierarsenault commented:
Thanks @gjohansson-ST,
- Tweaked the unique ID,
- Add mock to test values
- Moved the CONSTANT to (what I think) is the correct place in the class

marcolivierarsenault commented:
@gjohansson-ST or @frenck 

Any chance we can move forward with this PR. Thanks

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,97 @@
+class EcobeeVentilatorMinTime(NumberEntity):
+    """A number class, representing min time  for an ecobee thermostat with ventilator attached."""
+
+    VENTILATOR_MIN_VALUE = 0
+    VENTILATOR_MAX_VALUE = 60
+    VENTILATOR_STEP = 5
+    VENTILATOR_MEASUREMENT_UNIT = UnitOfTime.MINUTES
```
```suggestion
    _attr_native_min_value = 0
    _attr_native_max_value = 60
    _attr_native_step = 5
    _attr_native_unit_of_measurement = UnitOfTime.MINUTES
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,97 @@
+            name=self.thermostat["name"],
+        )
+        self._attr_unique_id = f'{self.thermostat["identifier"]}_ventilator_{mode}'
+        self._attr_native_min_value = self.VENTILATOR_MIN_VALUE
+        self._attr_native_max_value = self.VENTILATOR_MAX_VALUE
+        self._attr_native_step = self.VENTILATOR_STEP
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,97 @@
+        self._attr_native_max_value = self.VENTILATOR_MAX_VALUE
+        self._attr_native_step = self.VENTILATOR_STEP
+        self._attr_native_value = self.thermostat["settings"][ecobee_setting_key]
+        self._attr_native_unit_of_measurement = self.VENTILATOR_MEASUREMENT_UNIT
```
```suggestion
```

gjohansson-ST reviewed (changes requested):
Should be set at class

marcolivierarsenault committed:
add class variable

marcolivierarsenault commented:
@gjohansson-ST Applied your changes

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+
+    async def async_update(self):
+        """Get the latest state from the thermostat."""
+        await self.data.update()
```
This should be converted to use `DataUpdateCoordinator`

https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities

marcolivierarsenault replied:
as discuss, will be push to future pr

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+        self.ecobee_setting_key = ecobee_setting_key
+        self.set_func = set_func
+        self._attr_name = f"Ventilator min time {mode}"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self.thermostat["identifier"])},
+            manufacturer=MANUFACTURER,
+            model=ECOBEE_MODEL_TO_NAME.get(self.thermostat["modelNumber"]),
+            name=self.thermostat["name"],
+        )
```
Please move this into a base entity class so we can eventually get rid of the multiple ways `DeviceInfo` is created in each platform

bdraco replied:
I'd probably do another PR first to clean this up by creating the base entity class and than rebase this one on top of it once its merged which will make it much smaller

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,93 @@
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
+        self._attr_native_value = self.thermostat["settings"][self.ecobee_setting_key]
+
+    @property
+    def available(self):
+        """Return if device is available."""
+        return self.thermostat["runtime"]["connected"]
```
It looks like this could be moved into a base entity class as well

```
binary_sensor.py:        return thermostat["runtime"]["connected"]
climate.py:        return self.thermostat["runtime"]["connected"]
climate.py:        """Return true if humidifier connected to thermostat and set to manual/on mode."""
humidifier.py:        return self.thermostat["runtime"]["connected"]
number.py:        return self.thermostat["runtime"]["connected"]
sensor.py:        return thermostat["runtime"]["connected"]
```

bdraco commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

marcolivierarsenault commented:
@bdraco Thanks for the suggestion, 

Do you mind if we flip the PR order, ship this one and then create the base class for the entity platform, I rather have a small amount of PR open at the same time. (That being said, want to follow community guidelines, so if you think we absolutely need  the refactor of the entire Ecobee integration first, we can go that route too)

marcolivierarsenault committed:
add code owner to ecobee

bdraco commented:
> @bdraco Thanks for the suggestion, 
> 
> 
> 
> Do you mind if we flip the PR order, ship this one and then create the base class for the entity platform, I rather have a small amount of PR open at the same time. (That being said, want to follow community guidelines, so if you think we absolutely need  the refactor of the entire Ecobee integration first, we can go that route too)

You can split out the base class into an entity.py in this PR and only use it for your new entity, and than do future PRs to refactor each platform to use it.

marcolivierarsenault commented:
> > @bdraco Thanks for the suggestion,
> > Do you mind if we flip the PR order, ship this one and then create the base class for the entity platform, I rather have a small amount of PR open at the same time. (That being said, want to follow community guidelines, so if you think we absolutely need  the refactor of the entire Ecobee integration first, we can go that route too)
> 
> You can split out the base class into an entity.py in this PR and only use it for your new entity, and than do future PRs to refactor each platform to use it.

ooo smart, I like that. Will do this in the next 48h

marcolivierarsenault commented:
It will take me a bit more time than 72h, I will turn into draft PR and move back when ready

marcolivierarsenault commented:
@bdraco kind of hitting a snag here. 

Looks like the a previous dev of the Ecobee integration did hijack the `hass.data[DOMAIN]` to store a custom object. Looks like this is preventing to add a coordinator to the `hass.data.setdefault(DOMAIN, {})[entry.entry_id]` . 

Appart from a massive re-work of the integration, I am not sure how to fix this.

Do you have any recommendation on the strategy here?

bdraco commented:
> @bdraco kind of hitting a snag here.
> 
> Looks like the a previous dev of the Ecobee integration did hijack the `hass.data[DOMAIN]` to store a custom object. Looks like this is preventing to add a coordinator to the `hass.data.setdefault(DOMAIN, {})[entry.entry_id]` .

No need to do the `DataUpdateCoordinator` in this PR, it can be done in a followup.  
> 
> https://github.com/home-assistant/core/blob/4d215e573c79f7256035752a1d322d187a1e6fd4/homeassistant/components/ecobee/__init__.py#L72
> 
> Appart from a massive re-work of the integration, I am not sure how to fix this.
> 
> Do you have any recommendation on the strategy here?

I'd store a dataclass in there instead

example https://github.com/home-assistant/core/blob/dev/homeassistant/components/lookin/models.py

marcolivierarsenault committed:
adding base entity for ecobee devices

marcolivierarsenault commented:
@bdraco I did add a base class, looks to be working fine. 

The only issue is I don't have unit test for it (which I am not super comfortable) After looking through the code, looks like base class are rarely tested, is that expected?

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,34 @@
+        """Initiate base methods for Ecobee entities."""
+
+        self.thermostat = thermostat
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the device info."""
+        return DeviceInfo(
```
```suggestion
       self._attr_device_info = DeviceInfo(
```

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,34 @@
+        )
+
+    @property
+    def available(self):
```
```suggestion
    def available(self) -> bool:
```

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,34 @@
+class EcobeeBaseEntity:
+    """Base methods for Ecobee entities."""
+
+    def __init__(self, thermostat) -> None:
```
```suggestion
    def __init__(self, thermostat: dict[str, Any]) -> None:
```

bdraco commented on tests/components/ecobee/test_number.py:
```diff
@@ -0,0 +1,117 @@
+    """Test the number can set min time home."""
+    target_value = 40
+    with patch(
+        "pyecobee.Ecobee.set_ventilator_min_on_time_home"
```
```suggestion
        "homeassistant.components.ecobee.Ecobee.set_ventilator_min_on_time_home"
```

We should patch the object where its imported

bdraco commented:
> The only issue is I don't have unit test for it (which I am not super comfortable) After looking through the code, looks like base class are rarely tested, is that expected?

We shouldn't test the base class directly: Make sure to not interact with any integration details in tests of integrations. Following this pattern will make the tests more robust for changes in the integration.

https://developers.home-assistant.io/docs/development_testing?_highlight=test#writing-tests-for-integrations

marcolivierarsenault committed:
add comments

marcolivierarsenault commented:
@bdraco applied your comments. 
Thanks a lot for all the help

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,84 @@
+
+        self.data = data
+        self.thermostat_index = thermostat_index
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
```
```suggestion
        self.thermostat = data.ecobee.get_thermostat(self.thermostat_index)
```

bdraco committed:
Update homeassistant/components/ecobee/number.py

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,84 @@
+
+        self.data = data
+        self.thermostat_index = thermostat_index
+        self.thermostat = data.ecobee.get_thermostat(self.thermostat_index)
```
```suggestion
```

This is set in the base class

bdraco committed:
Update homeassistant/components/ecobee/number.py

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,31 @@
+
+        self.thermostat = thermostat
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self.thermostat["identifier"])},
+            manufacturer=MANUFACTURER,
+            model=ECOBEE_MODEL_TO_NAME.get(self.thermostat["modelNumber"]),
```
```suggestion
            identifiers={(DOMAIN, thermostat["identifier"])},
            manufacturer=MANUFACTURER,
            model=ECOBEE_MODEL_TO_NAME.get(thermostat["modelNumber"]),
```

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,83 @@
+    async def async_update(self):
+        """Get the latest state from the thermostat."""
+        await self.data.update()
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
```
Is this needed? Does the lib recreate the object every time?

marcolivierarsenault replied:
Yes it is needed, without it is does not update

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,83 @@
+    data = hass.data[DOMAIN]
+    entities = []
+    _LOGGER.debug("Adding min time ventilators numbers (if present)")
+    for index in range(len(data.ecobee.thermostats)):
```
Can you use `enumerate` here?

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,31 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class EcobeeBaseEntity:
```
```suggestion
class EcobeeBaseEntity(Entity):
```

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,31 @@
+import logging
+from typing import Any
+
+from homeassistant.helpers.entity import DeviceInfo
```
```suggestion
from homeassistant.helpers.entity import DeviceInfo, Entity
```

bdraco commented on homeassistant/components/ecobee/entity.py:
```diff
@@ -0,0 +1,31 @@
+            identifiers={(DOMAIN, self.thermostat["identifier"])},
+            manufacturer=MANUFACTURER,
+            model=ECOBEE_MODEL_TO_NAME.get(self.thermostat["modelNumber"]),
+            name=self.thermostat["name"],
```
```suggestion
            name=thermostat["name"],
```

marcolivierarsenault committed:
add base entity class

marcolivierarsenault commented:
@bdraco Comment reviewed, 
The only one not applied is about `self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)` which is required for the value to be updated

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,82 @@
+    for index, thermostat in enumerate(data.ecobee.thermostats):
+        if thermostat["settings"]["ventilatorType"] != "none":
+            _LOGGER.debug("Adding %s's ventilator min times number", thermostat["name"])
+            entities.append(
+                EcobeeVentilatorMinTime(
+                    data,
+                    index,
+                    "home",
+                    "ventilatorMinOnTimeHome",
+                    data.ecobee.set_ventilator_min_on_time_home,
+                )
+            )
+
+            entities.append(
+                EcobeeVentilatorMinTime(
+                    data,
+                    index,
+                    "away",
+                    "ventilatorMinOnTimeAway",
+                    data.ecobee.set_ventilator_min_on_time_away,
+                )
+            )
```
Since I have a feeling there is going to be more of these coming we should use `NumberEntityDescription` and create a `set_fn`

Example in homeassistant/components/deconz/number.py

marcolivierarsenault replied:
There could only be 2 type of `EcobeeVentilatorMinTime` Home and Away. (This is the way the integration works) 

If we add other type of numbers, I can't garante that they will require the same field types. 

Do you still think it is worth doing a `EcobeeVentilatorMinTime`

bdraco replied:
> If we add other type of numbers, I can't garante that they will require the same field types.
> 
> Do you still think it is worth doing a `EcobeeVentilatorMinTime`

Yes

bdraco replied:
Thanks

bdraco commented:
This looks good.  I've got one last (hopefully) piece of feedback here https://github.com/home-assistant/core/pull/83645/files#r1084676306

marcolivierarsenault committed:
add NumberEntityDescription

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,108 @@
+    entities = []
+    _LOGGER.debug("Adding min time ventilators numbers (if present)")
+    for index, thermostat in enumerate(data.ecobee.thermostats):
+        if thermostat["settings"]["ventilatorType"] != "none":
+            _LOGGER.debug("Adding %s's ventilator min times number", thermostat["name"])
+            for numbers in VENTILATOR_NUMBERS:
+                entities.append(EcobeeVentilatorMinTime(data, index, numbers))
```
```suggestion
        if thermostat["settings"]["ventilatorType"] == "none":
            continue
        _LOGGER.debug("Adding %s's ventilator min times number", thermostat["name"])
        for numbers in VENTILATOR_NUMBERS:
            entities.append(EcobeeVentilatorMinTime(data, index, numbers))
```

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,108 @@
+        self.ecobee_setting_key = description.ecobee_setting_key
+        self.set_fn = description.set_fn
+        self._attr_name = f"Ventilator min time {description.mode}"
+        self._attr_unique_id = (
+            f'{self.thermostat["identifier"]}_ventilator_{description.mode}'
+        )
+        self._attr_native_value = self.thermostat["settings"][
+            description.ecobee_setting_key
+        ]
```
Since we are stuck with a `dict` return type from the lib, in a future PR I'd make something like this in the base entity class

```
@property
def settings(self) -> dict[str, Any]:
    """Return ecobee settings."""
    return self.thermostat["settings"]
```

so you could do
```python
self._attr_native_value = self.settings[description.ecobee_setting_key]
```

marcolivierarsenault replied:
What is the reason we want to get this instead?

bdraco replied:
It more readable than this pattern over and over again:
```python
        self._attr_native_value = self.thermostat["settings"][
            description.ecobee_setting_key
        ]
```

bdraco commented:
I think we just have some test cleanup left and this should be good to merge.

I'll push it to my production system and do functional testing in a bit

marcolivierarsenault commented on tests/components/ecobee/test_number.py:
```diff
@@ -0,0 +1,117 @@
+        await hass.async_block_till_done()
+        mock_set_min_home_time.assert_called_once_with(THERMOSTAT_ID, target_value)
+
+    data = hass.data[ECOBEE_DOMAIN]
```
Not sure how to do this one tbh

bdraco replied:
https://github.com/home-assistant/core/pull/86100#discussion_r1083653828

marcolivierarsenault replied:
thanks

marcolivierarsenault replied:
Still looking, actually do we really need the second part of those test, 

```python
data = hass.data[ECOBEE_DOMAIN]

    monkeypatch.setitem(
        data.ecobee.get_thermostat(THERMOSTAT_ID)["settings"],
        "ventilatorMinOnTimeHome",
        target_value,
    )

    with patch("homeassistant.components.ecobee.Ecobee.update"):
        async_fire_time_changed(
            hass,
            dt.utcnow() + timedelta(minutes=5),
        )
        await hass.async_block_till_done()

    state1 = hass.states.get(VENTILATOR_MIN_HOME_ID)
    assert state1.state == str(target_value)
```

Looks like we are mocking the value we are looking for, looks very manual. If integration stop publishing to this we would not catch it with this test.

marcolivierarsenault replied:
Issue I have with patching this part is I need to patch the dictionary for the response because the service itself is already tested. 

```python 
target_value = 0
    with patch(
        "pyecobee.Ecobee.set_ventilator_min_on_time_away"
    ) as mock_set_min_away_time:
        await setup_platform(hass, DOMAIN)

        await hass.services.async_call(
            DOMAIN,
            SERVICE_SET_VALUE,
            {ATTR_ENTITY_ID: VENTILATOR_MIN_AWAY_ID, ATTR_VALUE: target_value},
            blocking=True,
        )
        await hass.async_block_till_done()
        mock_set_min_away_time.assert_called_once_with(THERMOSTAT_ID, target_value)
```

marcolivierarsenault replied:
@bdraco what do you think about this ☝️

bdraco replied:
That looks OK to me. I don't see any of the integration details being patched.

I'd probably change the `pyecobee.Ecobee.set_ventilator_min_on_time_away` patch target to be more exact (patch where its imported if possible), or store the patched object and modify it later instead

marcolivierarsenault committed:
reverse if logic

marcolivierarsenault committed:
remove hass accessing tests

marcolivierarsenault committed:
add data to fixture to include test with no add_ecobee_ventilator

bdraco committed:
Merge branch 'dev' into add_ecobee_ventilator

gjohansson-ST commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,109 @@
+        self.thermostat_index = thermostat_index
+        self.ecobee_setting_key = description.ecobee_setting_key
+        self.set_fn = description.set_fn
+        self._attr_name = f"Ventilator min time {description.mode}"
```
```suggestion
        self._attr_name = f"Ventilator min time {description.name}"
```

gjohansson-ST reviewed (changes requested):
I think `mode` should be replaced by the already there `name`.

marcolivierarsenault commented:
@gjohansson-ST I am good doing this, that being said, "home" and "away" are technically not the name of this entity. It would absolutely work but I am afraid it might get more confusing, because the entity name will be different aka `f"Ventilator min time {description.mode}"` 

Thoughts?

gjohansson-ST commented:
> @gjohansson-ST I am good doing this, that being said, "home" and "away" are technically not the name of this entity. It would absolutely work but I am afraid it might get more confusing, because the entity name will be different aka `f"Ventilator min time {description.mode}"`
> 
> Thoughts?

Name would be exactly as you have it now. I just ask you to remove the not needed `mode` key and replace with the built-in `name` as that's what you are using it for. Name would still be "ecobee .... home" for example.

marcolivierarsenault committed:
Merge branch 'dev' into add_ecobee_ventilator

marcolivierarsenault committed:
chage name from mode

marcolivierarsenault commented:
@gjohansson-ST applied your comments

bdraco commented:
@gjohansson-ST Thanks for the additional review

gjohansson-ST reviewed (approved):
LGTM 👍

bdraco committed:
Merge branch 'dev' into add_ecobee_ventilator

bdraco commented on homeassistant/components/ecobee/number.py:
```diff
@@ -0,0 +1,108 @@
+    async def async_update(self) -> None:
+        """Get the latest state from the thermostat."""
+        await self.data.update()
+        self.thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
```
I was thinking about how to avoid this.  

I think we could make thermostat a property that wrapped this code and than we wouldn't need to set it in init

marcolivierarsenault replied:
Would making it a property make it available to consume by the user? (I don't think we want that)

bdraco replied:
No, the user doesn't have direct access to the entity code

marcolivierarsenault replied:
Something like that ?? 
```python 
@property
    def thermostat_data(self):
        """get the latest cache data from the thermostat"""
        return self.data.ecobee.get_thermostat(self.thermostat_index)

    async def async_update(self) -> None:
        """Get the latest state from the thermostat."""
        await self.data.update()
        self._attr_native_value = self.thermostat_data["settings"][
            self.ecobee_setting_key
        ]   
```

bdraco replied:
I was thinking something like this. It should cherry-pick cleanly, but since I don't have an ecobee that ventilates I can't test it.

https://github.com/bdraco/home-assistant/commit/74f1bcfdc80d4dcab5f2547b81ec32035a223ab4

marcolivierarsenault replied:
Ok I cherry picked, and tested. It works perfectly. For some reason Github is not adding it to the PR, but it is in my fork,. https://github.com/marcolivierarsenault/core/commits/add_ecobee_ventilator 

I guess, GH will sync later

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
Merge remote-tracking branch 'marcolivierarsenault/add_ecobee_ventilator' into add_ecobee_ventilator

bdraco committed:
Merge branch 'dev' into add_ecobee_ventilator

bdraco committed:
Merge remote-tracking branch 'marcolivierarsenault/add_ecobee_ventilator' into add_ecobee_ventilator

marcolivierarsenault commented:
That is interesting, did not know this trick close/open...  Bug or intentional from GH?

bdraco commented:
> That is interesting, did not know this trick close/open... Bug or intentional from GH?

Something was busted with github. Close open is the usual trick to restart the CI and force resync

bdraco commented:
Thanks @marcolivierarsenault
